packages/@react-native-windows/codegen/Document.md

```diff
@@ -0,0 +1,386 @@
+
+Calling `MakeModuleProvider<T>` or `MakeTurboModuleProvider<T>` returns a `ReactModuleProvider` from a `REACT_MODULE` decorated struct `T`.
+`ReactModuleProvider` is a delegate, by passing a `IReactModuleBuilder` object as an argument, all reflectable members in the `REACT_MODULE` decorated struct will be enumerated and passed to `IReactModuleBuilder`.
+Your job is to implement `IReactModuleBuilder` to collect all members of a struct.
+
+### IReactModuleBuilder
+
+Here are all members in `IReactModuleBuilder`:
+
+- `void AddInitializer(InitializerDelegate)`.
+`InitializerDelegate` is a delegate taking `IReactContext` as an argument.
+A `REACT_MODULE` decorated struct expects all `InitializerDelegate` to be called before calling any other members.
+It is `IReactModuleBuilder`'s responsibility to prepare the `IReactContext` object.
+
+- `AddConstantProvider(ConstantProviderDelegate)`.
+`ConstantProviderDelegate` is a delegate taking `IJSValueWriter` as an argument.
+A `REACT_MODULE` decorated struct could provide multiple `ConstantProviderDelegate`,
+each delegate writes multiple members to the `IJSValueWriter`.
+In order to properly accept these members,
+delegates must be called between `IJSValueWriter::WriteObjectBegin` and `IJSValueWriter::WriteObjectEnd`,
+after that `IJSValueWriter` stores an object with all constant values as its fields.
+
+- `AddMethod(String, MethodReturnType, MethodDelegate)`.
+`MethodDelegate` represents an asynchronized method. The return value and the exception is returned by delegates in the 3rd and 4th arguments.
+`MethodReturnType` represents how this function looks like in JavaScript.
+  - `MethodReturnType::Callback`. The last argument is a callback to accept the return value, other n-1 arguments are passed to the delegate in `IJSValueReader`.
+  - `MethodReturnType::TwoCallbacks`. The last two arguments are callbacks for the return value and the exception, other n-2 arguments are passed to the delegate in `IJSValueReader`.
+  - `MethodReturnType::Void`. All arguments are passed to the delegate in `IJSValueReader`. This delegate do not offer return value or exception.
+  - `MethodReturnType::Promise`. All arguments are passed to the delegate in `IJSValueReader`.
+
+- `AddSyncMethod(String, SyncMethodDelegate)`.
+`SyncMethodDelegate` represents a synchronized method,
+taking `IJSValueReader` and `IJSValueWriter` as its arguments.
+The delegate will read all arguments from `IJSValueReader` as an array,
+and write the return value to `IJSValueWriter`.
+
+### IJSValueReader and IJSValueWriter
+
+Serialization and deserialization between C++ objects and `IJSValueReader/Writer` happens automatically.
+You should implement `IJSValueReader` and `IJSValueWriter` to pass arguments and read return values.
+
+An `IJSValueReader` or `IJSValueWriter` could access objects in JSON plain text, JSI objects or whatever is available in the other side of the platform.
+
+If you only need an abstract representation of C++ values,
```

Comment:
>If [](http://example.com/codeflow?start=0&length=2)

It is worth to mention that to serialize C++ values with help of the IJSValueReader and IJSValueWriter, developers should typically create ReadValue or WriteValue function overload for these types. The Microsoft.ReactNative.Cxx already offers them for the most common types.
The JSValue could be used to assist implementation of a complex ReadValue by de-serializing the value first into this intermediate represnetaiton.